### PR TITLE
Reduce minimum gas price to 100 MWEI

### LINKED ISSUE #2404: Reduce minimum gas price
Byzantium has somehow dramatically dropped commonly mined gas prices.

We should be able to safely reduce our enforced minimum gas price. First need to confirm our current gwei input handles decimals gracefully. Paging @Zanibas

### PR DESCRIPTION
**Pull request by danfinlay**
Also now representing gas price in MWEI, because our current bn inputs actually aren't tolerant of minimums less than 1 unit.

Fixes #2404.

screen shot 2017-10-24 at 3 07 48 pm
screen shot 2017-10-24 at 3 07 39 pm

### CODE REVIEW

**Review by 2-am-zzz — CHANGES_REQUESTED** [review]
can we remove the output css from this PR?

File: ui/app/components/pending-tx.js [hunk 1]
```diff
@@ -197,10 +197,10 @@ PendingTx.prototype.render = function () {
                 h(BNInput, {
                   name: 'Gas Price',
                   value: gasPriceBn,
-                  precision: 9,
-                  scale: 9,
-                  suffix: 'GWEI',
-                  min: MIN_GAS_PRICE_GWEI_BN.toString(10),
+                  precision: 6,
+                  scale: 6,
+                  suffix: 'MWEI',
```
Comment (Zanibas) [inline ui/app/components/pending-tx.js]: I disagree that we should be using MWEI units here--the general accepted units for gas manipulation is GWEI.

**Review by 2-am-zzz — DISMISSED** [review]
Failing tests as well for some reason?

File: ui/app/components/pending-tx.js (hunk 1 above)
Comment (danfinlay) [reply ui/app/components/pending-tx.js]: You think it’s better to leave the min gas price at 1 gwei than convert to mwei?

Comment (kumavis) [reply ui/app/components/pending-tx.js]: How much effort to add decimal support?

Comment (danfinlay) [reply ui/app/components/pending-tx.js]: I think a lot. I remember the amount of time spent on this one, to get it as stable as it is. Considering this directly affects funds spent, I don’t think it should be changed lightly.

**Comment by Zanibas** [comment]
How about these apples?

File: ui/app/components/pending-tx.js [hunk 2]
```diff
@@ -200,7 +200,7 @@ PendingTx.prototype.render = function () {
                   precision: 9,
                   scale: 9,
                   suffix: 'GWEI',
-                  min: MIN_GAS_PRICE_GWEI_BN.toString(10),
+                  min: MIN_GAS_PRICE_BN.toString(10),
```
Comment (Zanibas) [inline ui/app/components/pending-tx.js]: Just noticed, but we don't need the toString here!

**Comment by danfinlay** [comment]
Confirmed this seems to work. Sent a 0.1 gwei transaction, [visible here](https://rinkeby.etherscan.io/tx/0x097099327c7465e958a1e787f172a91a9e8038063a534e39874214a07f894dc8).

File: ui/app/components/bn-as-decimal-input.js [hunk 3]
```diff
@@ -128,15 +130,15 @@ BnAsDecimalInput.prototype.updateValidity = function (event) {
 }
 
 BnAsDecimalInput.prototype.constructWarning = function () {
-  const { name, min, max } = this.props
+  const { name, min, max, scale } = this.props
   let message = name ? name + ' ' : ''
 
   if (min && max) {
-    message += `must be greater than or equal to  ${min} and less than or equal to ${max}.`
+    message += `must be greater than or equal to  ${this.downsize(min.toString(10), scale)} and less than or equal to ${this.downsize(max.toString(10), scale)}.`
```
Comment (kumavis) [inline ui/app/components/bn-as-decimal-input.js]: can we use the `newMin` and `newMax` here?

File: ui/app/components/pending-tx.js [hunk 4]
```diff
@@ -57,7 +57,7 @@ PendingTx.prototype.render = function () {
   const safeGasLimit = safeGasLimitBN.toString(10)
 
   // Gas Price
-  const gasPrice = txParams.gasPrice || MIN_GAS_PRICE_BN.toString(16)
+  const gasPrice = txParams.gasPrice || MIN_GAS_PRICE_BN.toString(10)
```
Comment (kumavis) [inline ui/app/components/pending-tx.js]: why is the decimal `toString(10)` going into the `hexToBn`?

File: ui/app/components/pending-tx.js [hunk 5]
```diff
@@ -15,9 +15,9 @@ const addressSummary = util.addressSummary
 const nameForAddress = require('../../lib/contract-namer')
 const BNInput = require('./bn-as-decimal-input')
 
-const MIN_GAS_PRICE_GWEI_BN = new BN(1)
-const GWEI_FACTOR = new BN(1e9)
-const MIN_GAS_PRICE_BN = MIN_GAS_PRICE_GWEI_BN.mul(GWEI_FACTOR)
+const MIN_GAS_PRICE_MWEI_BN = new BN('100')
+const MWEI_FACTOR = new BN(1e6)
+const MIN_GAS_PRICE_BN = MIN_GAS_PRICE_MWEI_BN.mul(MWEI_FACTOR)
```
Comment (kumavis) [inline ui/app/components/pending-tx.js]: has the display changed to `MWEI`?

Comment (Zanibas) [reply ui/app/components/pending-tx.js]: This is only an internal change--the units reflected front-side are still in GWEI due to the scale/precision.

File: ui/app/components/pending-tx.js (hunk 4 above)
Comment (Zanibas) [reply ui/app/components/pending-tx.js]: Will fix

File: ui/app/components/bn-as-decimal-input.js (hunk 3 above)
Comment (Zanibas) [reply ui/app/components/bn-as-decimal-input.js]: We can, but I just did it here to continue to draw from props only rather other sources.

Comment (kumavis) [reply ui/app/components/bn-as-decimal-input.js]: right, ok maybe just move the calculation to the top of the fn
or dont, w/e

File: ui/app/components/pending-tx.js (hunk 5 above)
Comment (kumavis) [reply ui/app/components/pending-tx.js]: if this is only internal, why not go all the way down to wei?

Comment (Zanibas) [reply ui/app/components/pending-tx.js]: We can--it was just a preference I suppose to use the closest metric unit. Can change.

File: ui/app/components/bn-as-decimal-input.js (hunk 3 above)
Comment (Zanibas) [reply ui/app/components/bn-as-decimal-input.js]: :+1:

File: ui/app/components/pending-tx.js (hunk 5 above)
Comment (danfinlay) [reply ui/app/components/pending-tx.js]: No, display is showing Gwei.

**Comment by kumavis** [comment]
what happens when you put a lot of decimals? more decimals than `mwei` precision allows?

**Review by kumavis — APPROVED** [review]

**Comment by danfinlay** [comment]
@kumavis I tried, and the result was:

As I typed:

0.01, is fine
0.001, shows minimum wei error
As I kept adding zeroes, continued showing error, up to minimum precision:
0.0000001

Then the next zero popped it all to a zero, still showing the error.
0

This is slightly strange behavior, but appears totally safe.

**Comment by kumavis** [comment]
hmm i guess i meant more like `1.001`
but sounds like `1.0000001` -> `1`

**Comment by danfinlay** [comment]
When I tried typing `1.000000000001`, it was fine until I hit the final `1`, at which point the field turned to `1`, like you say.

**Comment by Zanibas** [comment]
It's a minor user inconvenience but never prohibits anyone from using wei precision.

**Pull request merged**